libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs

```diff
@@ -0,0 +1,393 @@
+        }
+
+        public HttpClient HttpClient { get; set; }
+
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (dc == null)
+            {
+                throw new ArgumentNullException(nameof(dc));
+            }
+
+            if (dc.Context?.Activity?.Type != ActivityTypes.Message)
+            {
+                return EndOfTurn;
+            }
+
+            var dialogOptions = new QnAMakerDialogOptions()
+            {
+                QnAMakerOptions = await GetQnAMakerOptionsAsync(dc).ConfigureAwait(false),
+                ResponseOptions = await GetQnAResponseOptionsAsync(dc).ConfigureAwait(false)
+            };
+
+            if (options != null)
+            {
+                dialogOptions = ObjectPath.Assign<QnAMakerDialogOptions>(dialogOptions, options);
+            }
+
+            ObjectPath.SetPathValue(dc.ActiveDialog.State, Options, dialogOptions);
+
+            return await base.BeginDialogAsync(dc, dialogOptions, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected async virtual Task<IQnAMakerClient> GetQnAMakerClientAsync(DialogContext dc)
+        {
+            var qnaClient = dc.Context.TurnState.Get<IQnAMakerClient>();
+            if (qnaClient != null)
+            {
+                // return mock client
+                return qnaClient;
+            }
+
+            var endpoint = new QnAMakerEndpoint
+            {
+                EndpointKey = this.endpointKey,
+                Host = this.hostName,
+                KnowledgeBaseId = this.knowledgeBaseId
+            };
+            var options = await GetQnAMakerOptionsAsync(dc).ConfigureAwait(false);
+            return new QnAMaker(endpoint, options, HttpClient);
+        }
+
+        protected virtual Task<QnAMakerOptions> GetQnAMakerOptionsAsync(DialogContext dc)
+        {
+            return Task.FromResult(new QnAMakerOptions
+            {
+                ScoreThreshold = this.threshold,
+                StrictFilters = this.strictFilters,
+                Top = this.top
+            });
+        }
+
+        protected virtual Task<QnADialogResponseOptions> GetQnAResponseOptionsAsync(DialogContext dc)
+        {
+            return Task.FromResult(new QnADialogResponseOptions
+            {
+                NoAnswer = noAnswer,
+                ActiveLearningCardTitle = activeLearningCardTitle ?? DefaultCardTitle,
+                CardNoMatchText = cardNoMatchText ?? DefaultCardNoMatchText,
+                CardNoMatchResponse = cardNoMatchResponse
+            });
+        }
+
+        private async Task<DialogTurnResult> CallGenerateAnswerAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            var dialogOptions = ObjectPath.GetPathValue<QnAMakerDialogOptions>(stepContext.ActiveDialog.State, Options);
+
```

Comment:
Can we insert `dialogOptions.QnAMakerOptions.QnaId = 0` just after this line. 
This is one issue that we found with our code and we resolved it for samples.